### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by rdevon**
Various bug fixes and enhancements:

visualize no longer ignores errors.
Added STL options.
Added option to not average in GAN scores
Added option to ignore nonlinearity on last conv layer
Updated resnets (needs to be tested)
There was a cuda error with some nonlinearities (PReLU). Added a fix, but probably cleaner solution exists.

### CODE REVIEW

File: cortex/built_ins/datasets/torchvision_datasets.py [hunk 1]
```diff
@@ -33,23 +33,45 @@ def _handle_SVHN(self, Dataset, data_path, transform=None, **kwargs):
...
-            normalize,
-        ])
-        test_transform = transforms.Compose([
-            transforms.Resize(64),
-            transforms.ToTensor(),
-            normalize,
-        ])
+
+        if stl_center_crop:
+            tr_trans = transforms.CenterCrop(64)
```
Comment (ReyhaneAskari) [inline cortex/built_ins/datasets/torchvision_datasets.py]: Are we always cropping to 64?

File: cortex/built_ins/datasets/torchvision_datasets.py [hunk 2]
```diff
@@ -33,23 +33,45 @@ def _handle_SVHN(self, Dataset, data_path, transform=None, **kwargs):
...
+            te_trans = transforms.CenterCrop(64)
+        elif stl_resize_only:
+            tr_trans = transforms.Resize(64)
```
Comment (ReyhaneAskari) [inline cortex/built_ins/datasets/torchvision_datasets.py]: same for resize?

**Comment by rdevon** [comment]
Fixed the errors. They were related to both the visualization changes and that tests were being done on mixed cpu /gpu.

File: cortex/built_ins/networks/utils.py [hunk 3]
```diff
@@ -75,7 +75,13 @@ def apply_nonlinearity(x, nonlinearity, **nonlinearity_args):
         if isinstance(nonlinearity, str):
             nonlinearity = get_nonlinearity(nonlinearity)
         if callable(nonlinearity):
-            x = nonlinearity(x, **nonlinearity_args)
+            if isinstance(nonlinearity, nn.PReLU):
```
Comment (ReyhaneAskari) [inline cortex/built_ins/networks/utils.py:78]: Shouldn't Pytorch be handling this? But if there is an error, shouldn't we check whether `x` is on `cuda` ?

File: cortex/built_ins/networks/resnets.py [hunk 4]
```diff
@@ -260,34 +252,19 @@ def __init__(self, shape, dim_out=None, dim_h=64,
             resblock = ResBlock(dim_in, dim_out, dim_x, dim_y, f_size,
                                 resample='down', name=name,
                                 spectral_norm=spectral_norm, **layer_args)
-            models.add_module(name, resblock)
+            self.models.add_module(name, resblock)
 
             dim_x //= 2
             dim_y //= 2
 
+        dim_out__ = dim_out
```
Comment (ReyhaneAskari) [inline cortex/built_ins/networks/resnets.py]: what is `dim_out__`?

File: cortex/built_ins/networks/convnets.py [hunk 5]
```diff
@@ -78,10 +78,18 @@ def __init__(self, shape, dim_out=None, dim_h=64,
                     dim_out = dim_h[i]
                 else:
                     dim_out = dim_in * 2
+            conv_args = dict((k, v) for k, v in layer_args.items())
             name = 'conv_({}/{})_{}'.format(dim_in, dim_out, i + 1)
             self.models.add_module(
                 name, Conv2d(dim_in, dim_out, f_size, stride, pad, bias=False))
             dim_x, dim_y = self.next_size(dim_x, dim_y, f_size, stride, pad)
+
+            last = not((dim_x >= min_dim and dim_y >= min_dim) and
```
Comment (ReyhaneAskari) [inline cortex/built_ins/networks/convnets.py]: what is `last`?

File: cortex/built_ins/datasets/torchvision_datasets.py [hunk 6]
```diff
@@ -33,23 +33,45 @@ def _handle_SVHN(self, Dataset, data_path, transform=None, **kwargs):
...
+            train_transform = transforms.Compose([
+                transforms.RandomHorizontalFlip(),
+                transforms.ToTensor(),
+                normalize,
+            ])
+            test_transform = transforms.Compose([
+                transforms.ToTensor(),
+                normalize,
+            ])
+        else:
```
Comment (ReyhaneAskari) [inline cortex/built_ins/datasets/torchvision_datasets.py:51]: You could bring the upper `if`, `else` block down here.

**Review by ReyhaneAskari — COMMENTED** [review]
Better add some comments or use better names for some variables.

File: cortex/built_ins/datasets/torchvision_datasets.py (hunk 2 above)
Comment (rdevon) [reply cortex/built_ins/datasets/torchvision_datasets.py]: Not if you pass `stl_no_resize`

File: cortex/built_ins/networks/resnets.py (hunk 4 above)
Comment (rdevon) [reply cortex/built_ins/networks/resnets.py]: the output depth of the last resblock

File: cortex/built_ins/networks/convnets.py (hunk 5 above)
Comment (rdevon) [reply cortex/built_ins/networks/convnets.py]: last conv layer

File: cortex/built_ins/datasets/torchvision_datasets.py (hunk 6 above)
Comment (rdevon) [reply cortex/built_ins/datasets/torchvision_datasets.py:51]: done

File: cortex/built_ins/networks/resnets.py (hunk 4 above)
Comment (rdevon) [reply cortex/built_ins/networks/resnets.py]: I'll change the name of these things

File: cortex/built_ins/networks/utils.py (hunk 3 above)
Comment (rdevon) [reply cortex/built_ins/networks/utils.py:78]: Yes, changed to use `x.device`

File: cortex/built_ins/datasets/torchvision_datasets.py (hunk 1 above)
Comment (rdevon) [reply cortex/built_ins/datasets/torchvision_datasets.py]: (see below)

**Review by ReyhaneAskari — APPROVED** [review]

**Review by TiSU32 — COMMENTED** [review]
I looked at the changes and Reyhane's comments. I don't see anything wrong.

**Pull request merged**